rfcs/0042-local-test-framework.md

```diff
@@ -0,0 +1,272 @@
+
+```ocaml
+let rec pull_subscription_in_background ~logger ~network ~event_writer
+    ~subscription =
+  if not (Pipe.is_closed event_writer) then (
+    [%log spam] "Pulling StackDriver subscription" ;
+    let%bind log_entries =
+      Deferred.map (Subscription.pull ~logger subscription) ~f:Or_error.ok_exn
+    in
+    if List.length log_entries > 0 then
+      [%log spam] "Parsing events from $n logs"
+        ~metadata:[("n", `Int (List.length log_entries))]
+    else [%log spam] "No logs were pulled" ;
+    let%bind () =
+      Deferred.List.iter ~how:`Sequential log_entries ~f:(fun log_entry ->
+          log_entry
+          |> parse_event_from_log_entry ~network
+          |> Or_error.ok_exn
+          |> Pipe.write_without_pushback_if_open event_writer ;
+          Deferred.unit )
+    in
+    let%bind () = after (Time.Span.of_ms 10000.0) in
+    pull_subscription_in_background ~logger ~network ~event_writer
+      ~subscription )
+  else Deferred.unit
+```
+
+[https://github.com/MinaProtocol/mina/blob/67cc4205cc95138cf729a2f14b57b754f9e9204e/src/lib/integration_test_cloud_engine/stack_driver_log_engine.ml#L269](https://github.com/MinaProtocol/mina/blob/67cc4205cc95138cf729a2f14b57b754f9e9204e/src/lib/integration_test_cloud_engine/stack_driver_log_engine.ml#L269)
+
+A similar interface can be written for Docker-Swarm instead. By defining a `Service.pull` function with a given logger, we can leverage a lot of the work already done by modifying parts of the code where the log formats diverge. All logs can be specified to an output stream, such as stdout or a specified file by the user on their local system. 
+
+<img src="./res/local-test-integration-logging.png" alt="drawing" width="500"/>
+
+# Work Breakdown/Prio
+
+The following will be a work breakdown of what needs to be done to see this feature to completion:
+
+1. Implement the `Network_Config` interface to accept a network configuration and create a corresponding `docker-compose.json` file.
+2. Implement the `Network_manager` interface to take a corresponding `docker-compose.json` file and create a local swarm with the specified container configuration
+3. Implement functionality to simply log all container logs into a single stream (stdout or a file, maybe this can be specified in startup?)
+4. Implement filter on event functionality
+5. Ensure that current integration test specs are able to run on the local framework with success
+
+
+# Unresolved Questions
+
+- Is compiling a docker-compose file the right approach for scheduling the containers? The nice thing about using a docker-compose file is that all network management should be automatic.
+
+- Is using a different service for each type of node the best effective approach? Would it be better to launch all nodes under the same service in the docker-compose file?
+
+- Is polling each service and then aggregating those logs the best approach? Would it be better to do filtering before aggregating?
```

Comment:
I think the polling logs in the local integration test engine is not the correct approach.

We do a pull-based approach for the cloud engine, and that works out fine for us right now because we pre-filter the logs (so the volume we handle per poll is very low), but more importantly, because we are only running networks with snarks enabled. When snarks are enabled, the networks move pretty slow, so small delays in the order of 5-10 seconds are not a problem.

However, with the local engine, the goal is the run snarkless networks, which move extremely quickly by comparison. In that world, a 5-10 second delay in logs could break the test. Furthermore, the log volume per second of a snarkless network is *massive* by comparison due to the speed the network operates at. Because of this, I think we both need to pre-filter the logs in the docker containers (to bring the log volume the single-threaded test_executive needs to handle down to a fraction of the size), and we need to be reading those logs off of a pipe in a more push-based fashion.

Another advantage of using a push-based pipe approach here is that you don't need to deal with re-reading logs you already handled, which appears to be a problem you would need to solve with the docker swarm logging interface. You might be able to get by with asking docker swarm to only give you logs that were recorded since the last time you asked for logs, but you have some risk of missing logs in that approach and are relying on docker swarm's performance to quickly perform that query on it's log cache.

Reply:
Followed up with a restructuring of the proposed polling solution to use the push-based pipe approach.  👍